File: app11_swift_ios/SwiftGuardData/Sources/SwiftGuardData/Cards/ReferenceValidator.swift
import Foundation

/// SR-07-equivalent: owaspRefs/mitreRefs values are validated against
/// allowlists bundled as app resources before any card is written to
/// SwiftData — the same guard app09's Reference_Validator provides, since
/// every one of these values is curated content (PLAN.md §0.1), never
/// extracted from the raw YAML.
public struct ReferenceValidator: Sendable {
    private let owaspAllowlist: Set<String>
    private let mitreAllowlist: Set<String>

    public init(bundle: Bundle = .main) throws {
        owaspAllowlist = try Self.loadAllowlist(named: "ref-allowlists", key: "owasp_refs", bundle: bundle)
        mitreAllowlist = try Self.loadAllowlist(named: "mitre-atlas-allowlist", key: "mitre_refs", bundle: bundle)
    }

    public func assertOwaspRefsValid(_ refs: [String], cardId: String) throws {
        for ref in refs where !owaspAllowlist.contains(ref) {
            throw CardDecodeError.unknownReference(value: ref, field: "owasp_refs", cardId: cardId)
        }
    }

    public func assertMitreRefsValid(_ refs: [String], cardId: String) throws {
        for ref in refs where !mitreAllowlist.contains(ref) {
            throw CardDecodeError.unknownReference(value: ref, field: "mitre_refs", cardId: cardId)
        }
    }

    private static func loadAllowlist(named name: String, key: String, bundle: Bundle) throws -> Set<String> {
        guard let url = bundle.url(forResource: name, withExtension: "json") else {
            return []
        }
        let data = try Data(contentsOf: url)
        struct AllowlistFile: Decodable {
            let values: [String]
            enum CodingKeys: String, CodingKey {
                case owaspRefs = "owasp_refs"
                case mitreRefs = "mitre_refs"
            }
            init(from decoder: Decoder) throws {
                let container = try decoder.container(keyedBy: CodingKeys.self)
                if let owasp = try container.decodeIfPresent([String].self, forKey: .owaspRefs) {
                    values = owasp
                } else {
                    values = try container.decode([String].self, forKey: .mitreRefs)
                }
            }
        }
        return Set(try JSONDecoder().decode(AllowlistFile.self, from: data).values)
    }
}

File: app14_LUA_UNITY/backend/app/services/reference_validator.lua
local json = require("cjson")
local CardDecodeError = require("app.services.card_decode_error")

-- SR-07/D-06: every curated owasp_refs/mitre_refs value is validated against
-- a bundled allowlist before being written to the database — an unknown
-- reference aborts ingestion, mirroring app13_ruby_FastApi's ReferenceValidator.
local ReferenceValidator = {}
ReferenceValidator.__index = ReferenceValidator

local function read_json(path)
  local f = io.open(path, "r")
  if not f then return nil end
  local content = f:read("*a")
  f:close()
  return json.decode(content)
end

function ReferenceValidator.new(seeds_root)
  local self = setmetatable({}, ReferenceValidator)
  self.seeds_root = seeds_root

  local owasp = read_json(seeds_root .. "/ref-allowlists.json") or {}
  local mitre = read_json(seeds_root .. "/mitre-atlas-allowlist.json") or {}

  self.owasp_refs = {}
  for _, ref in ipairs(owasp.owasp_refs or {}) do self.owasp_refs[ref] = true end

  self.mitre_refs = {}
  for _, ref in ipairs(mitre.mitre_refs or {}) do self.mitre_refs[ref] = true end

  return self
end

function ReferenceValidator:assert_owasp_refs_valid(refs, card_id)
  for _, ref in ipairs(refs or {}) do
    if not self.owasp_refs[ref] then
      error(CardDecodeError.unknown_reference(ref, "owasp_refs", card_id))
    end
  end
end

function ReferenceValidator:assert_mitre_refs_valid(refs, card_id)
  for _, ref in ipairs(refs or {}) do
    if not self.mitre_refs[ref] then
      error(CardDecodeError.unknown_reference(ref, "mitre_refs", card_id))
    end
  end
end

return ReferenceValidator
